data/data/azure/vnet/main.tf

```
@@ -50,6 +50,15 @@ resource "azurerm_storage_account" "cluster" {
   account_replication_type        = "LRS"
   min_tls_version                 = contains(local.environments_with_min_tls_version, var.azure_environment) ? "TLS1_2" : null
   allow_nested_items_to_be_public = false
+
+  dynamic "network_rules" {
```

This can be done with a single network rule using multiple virtual network ID's rather than multiple network rules. This will also work if somehow there are no subnets. The issue here is figuring out how to get the public IP address of the machine doing the installation. There isn't a clean way of doing it. 

```suggestion
  network_rules {
    default_action             = "Deny"
    virtual_network_subnet_ids = concat(
      [for subnet in azurerm_subnet.master_subnet: subnet.id],
      [for subnet in azurerm_subnet.worker_subnet: subnet.id]
    )
    ip_rules                   = ["<Public IP Address>"]
    bypass                     = ["Logging", "AzureServices", "Metrics"]
  }

```

---

+1 It doesn't seem like we need the dynamic block here. There should only be two subnets, so I think you can use the local ID variables:
https://github.com/openshift/installer/blob/3d92c1e6dd1fa1588810ac57223261ff2c74a047/data/data/azure/vnet/common.tf#L35-L36

> The issue here is figuring out how to get the public IP address of the machine doing the installation. There isn't a clean way of doing it.

I'm not certain the installer host needs to access the storage account--at least directly. The storage account does contain boot diagnostic logs, but I assume those are accessed indirectly through azure. If the diagnostic log access is an issue, there has been discussion from the cloud team about moving those to an azure managed service account. There is benefit in doing that because we could then delete this storage account at bootstrap destroy. But that's probably beyond the scope of this bug fix.

---

It appears as though it is required to specify the IP or IP range of the machine doing the installation. It looks like we can dodge this bullet by using a private endpoint. That looks like the proper route to go here.

---

> It appears as though it is required to specify the IP or IP range of the machine doing the installation

I'm not following where or why it is required to get the installer host ip. Can you explain?

`ip_rules` is an optional field: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/storage_account_network_rules#ip_rules